README.md
jquerySimpleFAQ
===============

Simple jQuery plugin for creating FAQ interfaces

Features
--------

* __Simple to use.__ Use existing HTML data (ul > li) for source or use a JSON data array
* __Lots of options, but none required.__
* __Fast live searching using Quicksilver, including relevancy sorting (optionally).__
* __Add tags to FAQs to increase relevancy of searching.__
* __Events triggered for extra preocess handling if necessary.__


Basic Usage
-----------

With an HTML data source:
```html
<ul id='faqs'>
  <li>
    <!-- Note that you MUST have the "question", "answer", and "tags" classes on these nodes -->
    <p class='question'>This is a question?</p>
    <div class='answer'>This is the answer.</div>
    <p class='tags'>tags, help, searching</p>
  </li>
  ...
</ul>
```

```javascript
$('#faqs').simpleFAQ(); // Most simple form (all default options)
```

When using the 'data' option (JSON) and searching:

```html
<div><input type='text' id='faqSearch' /></div>
<ul id='faqs'></ul>
```

```javascript
$('#faqs').simpleFAQ({
  data: [
    {
      question: "This is a question?",
      answer: "This is the answer.",
      tags: "tags, help, searching" // OPTIONAL, useful for search scoring (and displaying if you wish)
    },
    ...
  ],
  allowSearch: true,
  searchNode: '#faqSearch'
});
```

See more examples at: http://jordankasper.com/jquery/faq/examples.php

Events
------

* _init.simpleFAQ_: The basic UI container and event handlers have been added to the document (or have been updated). You should only ever receive ONE of these events per page load. _(Arguments: event, `$.jk.SimpleFAQ` object)_
* _show.simpleFAQ_: Fired when an individual FAQ answer is shown. (Note: when searching for results, this will NOT fire when the FAQ is shown, only when the ANSWER is shown.) _(Arguments: event, `HTMLElement` faq)_
* _hide.simpleFAQ_: Fired when an individual FAQ answer is hidden (either explicitly by the user, or because another FAQ was shown and the `showOnlyOne` option is `true`. (Note: when searching for results, this will NOT fire when the FAQ is hidden because it is not in the result set, only when the ANSWER is hidden after being shown.) _(Arguments: event, `HTMLElement` faq)_
* _searchStart.simpleFAQ_: Fires when the user initiates a search. _(Arguments: event)_
* _searchEnd.simpleFAQ_: Fired after a search sction is complete (all results will be showing at this point). _(Arguments: event, array scores (NOTE: format will be: [ [score, faqNode], ...] ))_
* _sort.simpleFAQ_: Fired when the results of a search are sorted. _(Arguments: event, array scores (NOTE: format will be: [ [score, faqNode], ...] ))_


Event Example
-------------

```javascript
$('#faqs').bind('searchEnd.simpleFAQ', function(jQEvent, results) {
  if (results.length < 1) {
    // do something when there are no results?
  }
});
```

Options
-------

* _node_: `selector | HTMLElement | jQuery object` The node (or selector) to use for the FAQ UI. If not set, the current node selected by $(...).simpleFAQ(); will be used. _(default: `null`)_
* _data_: `array` The JSON data to use for the FAQs (will be added to any HTML FAQs). Format: `[ { question: "...", answer: "...", tags: "multiple, tags" /* OPTIONAL */ } ]` _(default: `null`)_
* _ns_: `string` Used before all assigned classes and as an event namespace. _(default: `"simpleFAQ"`)_
* _nodeType_: `string` The type of node to look for (or use with JSON data) for FAQs. _(default: `"li"`)_
* _questionClass_: `string` The class that all questions will have (either you have to give them this class, or use the plugin to build the list). _(default: `"question"`)_
* _answerClass_: `string` The class that all answers will have (either you have to give them this class, or use the plugin to build the list). _(default: `"answer"`)_
* _tagClass_: `string` The class for a node in the answer that contains tags specific to each answer. If this exists, it boosts the score for search terms that are in the tags. _(default: `"tags"`)_
* _showOnlyOne_: `boolean` If true, only one answer will be visible at a time (accordion style). _(default: `false`)_
* _changeHash_: `boolean` If true, the URL hash will be changed on each FAQ toggle, thus allowing for linking directly to a specific FAQ. _(default: `true`)_
* _slideSpeed_: `number | string` The speed to open and close FAQ answers. String values must be one of the three predefined speeds: "slow", "normal", or "fast"; numeric values are the number of milliseconds to run the animation (e.g. 1000). _(default: `500`)_

_Search Options_

* _allowSearch_: `boolean` If true, adds searching ability (must provide searchNode) _(default: `false`)_
* _searchNode_: `selector | HTMLElement | jQuery object` Only required if `allowSearch` is `true`; it is the element used for search input. NOTE: we use the `keyup` event, so this should be something that will emit that event correctly and can have a `value`! (Can be a node, jQuery object, or selector.) _(default: `null`)_
* _minSearchScore_: `number` The minimum score a FAQ must have in order to appear in search results. Depends on what search function youa re using, but quicksilver returns scores between 0 and 1 (although the final score can be more than 1 based on tag scoring). _(default: `0.5`)_
* _sortSearch_: `boolean` Whether or not to sort search results by score (descending). _(default: `false`)_
* _showAllOnEmpty_: `boolean` Should the plugin show all FAQs when there is no search input? _(default: `true`)_
* _caseSensitive_: `boolean` Whether or not the search is case sensitive. _(default: `false`)_
* _keyTimeout_: `number` A number of milliseconds to wait after a keyup event before initiating a search. Allows for better efficiency as a person is typing (live search principle). _(default: `400`)_
* _partialTagScore_: `number` What to increase the match score by when partial tags are matched (such as "sim" -> "simple") _(default: `0.1`)_
* _exactTagScore_: `number` What to increase the match score by when an exact tag is matched (such as "simple" -> "simple") _(default: `0.2`)_
* _score_: `function` A function used to score FAQ question and answer text (not tags). Should accept the full text and a single search token and return the score (generally a decimal number between 0 and 1, but this is up to you). _(default: `$.score` (Quicksilver scoring function))_


CHANGELOG

Version 1.0.0
 - Many changes which will BREAK existing usage. Look for (COMPATABILITY WARNING) messages below!
 - Moved everything to github
 - Changed minimum support to jQuery 1.7 (COMPATABILITY WARNING)
 - Reorganized files
 - Organized code into more OOP model (better for testing)
 - Added test suite
 - Added namespace to all events
 - Added various events
 - Changes to some class names (COMPATABILITY WARNING)
 - Various efficiencies (and removed some unnecessary code)
 - Removed forced ID on FAQ node (COMPATABILITY WARNING)
 - Change to structure of HTML when FAQs generated from JSON data (COMPATABILITY WARNING)
 - Removed ability to ignore words in scoring search results (COMPATABILITY WARNING)
 - The searchEnd event now returns 2-dimensional array with search scores and nodes (COMPATABILITY WARNING)
 - Added option to pass in a custom "score" function for FAQ text

Version 0.7
 - Minor bug fix (leftover console statement) and file size reduction

Version 0.6
 - Added direct FAQ jumping using URL anchors (thanks to Ian G. for the idea and starting code)

Version 0.5
 - Fixed bug where nodeTypes (usually "li") within a FAQ were selected as FAQs
   (In other words, you had a FAQ that was an LI, and then an LI inside that.)
 - Added audit for single root node in jQuery chain, otherwise we kind of break.

Version 0.4
 - Fixed issue with class on shown answers
   NOTE: The answer div no longer has the simpleFAQShowing class added!
         In your CSS you will need to use "li.simpleFAQShowing .answer".

Version 0.3
 - Added caseSensitive option for searches
 - Added list of words to ignore when searching/scoring
 - Added "searchStart" and "searchEnd" events
 - Added more classes to FAQs found by searching

Version 0.2
 - Added speed option (thanks to Ferenc Radius)
 - Minor other changes (size, comments/notes)

Version 0.1
 - Initial Release


src/jquery.simpleFAQ.js
/* Copyright (c) 2008 Jordan Kasper
 * Licensed under the MIT (http://www.opensource.org/licenses/mit-license.php)
 * Copyright notice and license must remain intact for legal use
 * Requires: jQuery 1.2+
 *           jQuery.quicksilver
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS 
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 * 
 * For more usage documentation and examples, visit:
 *  https://github.com/jakerella/jquerySimpleFAQ
 *  
 */
;(function($) {
  
  // ----------- Public methods ----------- //
  
  $.fn.simpleFAQ = function(o) {
    var n = this;
    if (n.length != 1) { return n; }
    n.addClass('simpleFAQList');
    
    // Set up options (and defaults)
    o = (o)?o:{};
    o = audit($.extend({}, $.fn.simpleFAQ.defaults, o));
    
    // Make sure node has an ID
    if (n.attr('id').length < 1) {
      n.attr('id', 'simpleFAQ_'+Math.floor(Math.random() * 1000));
    }
    
    var qc = o.questionClass;
    var ac = o.answerClass;
    var nt = o.nodeType;
    
    // Are we building the FAQs from data? (audited above)
    if (o.data != null) {
      var d = o.data;
      n.html('');
      for (var i=0, l=d.length; i<l; ++i) {
        n.append(
          "<"+nt+" class='simpleFAQ'>"+
          " <p class='"+qc+"'>"+d[i].question+"</p>"+
          " <div class='"+ac+"'>"+
          d[i].answer+
          "<p class='"+o.tagClass+"'>"+d[i].tags+"</p>"+
          "</div>"+
          "</"+nt+">"
        );
      }
    }
    
    // Cache all FAQ nodes (only first children)
    var faqs = $('#'+n.attr('id')+' > '+nt);
    
    // Show answers when question clicked
    faqs
      .find('.'+qc)
        .css('cursor', 'pointer')
        .hover(
          function () { $(this).addClass('simpleFAQHover'); },
          function() { $(this).removeClass('simpleFAQHover'); }
        )
        .bind('click.simpleFAQ', function(e) {
          var f = $(this).parent();
          if (o.showOnlyOne) {
            // Hide all others
            n.find(nt)
              .not(f)
                .find('.'+ac)
                  .slideUp(o.speed, function() {
                    $(this).parent()
                      .removeClass('simpleFAQShowing');
                  });
          }
          $(this)
            .siblings('.'+ac)
              .slideToggle(o.speed, function() {
                if ($(this).is(':visible')) {
                  f.addClass('simpleFAQShowing');
                  var h = f.attr('id');
                  if (h && h.length > 0) {
                    document.location.hash = escape(h);
                  }
                  n.trigger('show.simpleFAQ', [f[0]]);
                } else {
                  f.removeClass('simpleFAQShowing');
                }
              });
        });
    
    // Hide all answers by default
    faqs.find('.'+ac).hide();
    
    // Searching is enabled
    if (o.allowSearch) {
      // Helper for hiding FAQs when not in search results
      var hideFAQ = function(node) {
        $(node)
          .hide()
          .removeClass('simpleFAQResult')
          .find('.'+ac)
            .hide()
            .parent()
              .removeClass('simpleFAQShowing');
      }
      
      // create input node
      var sn = $(o.searchNode);
      if (sn.length > 0 && typeof $.score == 'function') {
        // Hide all FAQs, they'll be shown when found in a search
        hideFAQ(n.find(nt));
        
        var h;
        sn
          .append("<input type='text' id='simpleFAQSearch' />")
          .find('#simpleFAQSearch')
            .keyup(function(e) {
              clearTimeout(h);
              var sn = this;
              if (sn.value.length < 1) {
                hideFAQ(n.find(nt));
                return;
              }
              
              // add a slight delay to wait for more input
              h = setTimeout(function() {
                n.trigger('searchStart.simpleFAQ', []);
                // Score the input
                var sc = [];
                faqs.each(function(i) {
                  var f = $(this);
                  var tg = f.find('.'+o.tagClass).text();
                  tg = (o.caseSensitive)?tg:tg.toLowerCase();
                  var t = f.text();
                  t = (o.caseSensitive)?t:t.toLowerCase();
                  var q = getQuery(sn.value, o);
                  var s = 0;
                  
                  if (q.length > 0) {
                    s = $.score(t, q);
                    s += scoreTags(q, tg);
                  }
                  if (s > o.minSearchScore) {
                    sc.push([s, f]);
                  } else {
                    hideFAQ($(this));
                  }
                });
                
                if (o.sortSearch) {
                  // Sort results
                  sc.sort(function(a, b){
                    return b[0] - a[0];
                  });
                }
                
                // Show the relevant questions
                var r = [];
                $.each(sc, function() {
                  n.append(this[1].show().addClass('simpleFAQResult'));
                  r.push(this[1][0]);
                });
                
                n.trigger('sort.simpleFAQ', [r]);
                n.trigger('searchEnd.simpleFAQ', [r]);
                
              }, $.fn.simpleFAQ.keyTimeout);
            });
      }
    }
    
    var scoreTags = function(t, tags) {
      var s = 0;
      if (tags.length < 1) { return s; }
      var w = t.split(' ');
      for (var i=0, l=w.length; i<l; ++i) {
        if (w[i].length < 1) { continue; }
        if (tags.indexOf(w[i]) > -1) {
          s += $.fn.simpleFAQ.tagMatchScore;
        }
      }
      return s;
    }
    
    var getQuery = function(t, o) {
      var q = '';
      t = (o.caseSensitive)?t:t.toLowerCase();
      var ig = o.ignore;
      if (ig.length > 0) {
        var w = t.split(' ');
        for (var i=0; i<w.length; ++i) {
          if (w[i].length > 0) {
            if (typeof ig.indexOf == 'function') {
              if (ig.indexOf(w[i]) < 0) {
                q += w[i] + ' ';
              }
            } else {
              var f = false;
              for (var j=0; j<ig.length; ++j) {
                if (ig[j] == w[i]) {
                  f = true;
                  break;
                }
              }
              if (!f) { q += w[i] + ' '; }
            }
          }
        }
        if (q.length > 0) { q = q.substr(0, q.length-1); }
      } else {
        q = t;
      }
      return q;
    }
    
    // See if we have a starting FAQ and show it
    var h = document.location.hash;
    if (h && h.length > 0) {
      var fn = $(h);
      if (fn && fn.is('.simpleFAQList>*')) {
        fn.find('.'+qc).trigger('click.simpleFAQ');
      }
    }
    
    // Return original chain of nodes to continue jQuery chain
    return n;
  };
  
  // Defined outside simpleFAQ to allow for usage during construction
  var audit = function(o) {
    var d = o.data;
    if (!d || !d.length || typeof d.splice != 'function') {
      o.data = $.fn.simpleFAQ.defaults.data;
    }
    if (typeof o.nodeType != 'string') { o.nodeType = $.fn.simpleFAQ.defaults.nodeType; }
    if (typeof o.questionClass != 'string') { o.questionClass = $.fn.simpleFAQ.defaults.questionClass; }
    if (typeof o.answerClass != 'string') { o.answerClass = $.fn.simpleFAQ.defaults.answerClass; }
    if (typeof o.tagClass != 'string') { o.tagClass = $.fn.simpleFAQ.defaults.tagClass; }
    
    if (typeof o.showOnlyOne != 'boolean') { o.showOnlyOne = $.fn.simpleFAQ.defaults.showOnlyOne; }
    if (typeof o.allowSearch != 'boolean') { o.allowSearch = $.fn.simpleFAQ.defaults.allowSearch; }
    if (typeof o.minSearchScore != 'number') { o.minSearchScore = $.fn.simpleFAQ.defaults.minSearchScore; }
    if (typeof o.sortSearch != 'boolean') { o.sortSearch = $.fn.simpleFAQ.defaults.sortSearch; }
    if (typeof o.caseSensitive != 'boolean') { o.caseSensitive = $.fn.simpleFAQ.defaults.caseSensitive; }
    if (typeof o.speed != 'number') { o.speed = $.fn.simpleFAQ.defaults.speed; }
    
    var ig = o.ignore;
    if (!ig || !ig.length || typeof ig.splice != 'function') {
      o.ignore = $.fn.simpleFAQ.defaults.ignore;
    }
    
    return o;
  }
  
  
  // ----------- Static properties ----------- //
  
  $.fn.simpleFAQ.keyTimeout = 400;
  $.fn.simpleFAQ.tagMatchScore = 0.1;
  
  // options for simpleFAQ instances...
  $.fn.simpleFAQ.defaults = {
    data: null,                // Array If provided, this data is used as the FAQ data with each array entry being an object with 'question', 'answer', and 'tags' properties, this will be used to build the list
    nodeType: 'li',            // String The type of node to look for (and use) for FAQs
    questionClass: 'question', // String The class that all questions will have (either you have to give them this class, or use the plugin to build the list)
    answerClass: 'answer',     // String The class that all answers will have (either you have to give them this class, or use the plugin to build the list)
    tagClass: 'tags',          // String The class for a node in the answer that contains tags specific to each answer. If this exists, it boosts the score for search terms that are in the tags
    showOnlyOne: false,        // Boolean If true, only one answer will be visible at a time
    allowSearch: false,        // Boolean If true, adds a search box (must provide searchNode)
    searchNode: null,          // jQ Node  Only required if allowSearch is true; it is the container for the search box (should be a node, the jQuery object, or a selector)
    minSearchScore: 0,         // Number The minimum score a FAQ must have in order to appear in search results. Should be a number between 0 and 1 (Quicksilver score)
    sortSearch: false,         // Boolean Whether or not to sort search results
    caseSensitive: false,      // boolean Whether or not the search is case sensitive
    speed: 500,                // Number or String The speed to open and close FAQ answers. String values must be one of the three predefined speeds: "slow", "normal", or "fast"; numeric values are the number of milliseconds to run the animation (e.g. 1000).
    ignore: ['the', 'a', 'an', 'i', 'we', 'you', 'it', 'that', 'those', 'these', 'them', 'to', 'and', 'or', 'as', 'at', 'by', 'for', 'of', 'so']
                                // Array A list of words to ignore when searching
  };

})(jQuery);
